File: SW/FNTT/fntt.c
#include <stdio.h>
#include <math.h>

int reverse_bits(int number, int bit_length) {
    int reversed = 0;
    for (int i = 0; i < bit_length; i++) {
        if ((number >> i) & 1) {
            reversed |= 1 << (bit_length - 1 - i);
        }
    }
    return reversed;
}

int* cooley_tukey_ntt(int* out, int q, int* omegas, int n) {
    for (int i = 0; i < n; i++) {
        int rev_i = reverse_bits(i, (int)log2(n));
        if (rev_i > i) {
            out[i] ^= out[rev_i];
            out[rev_i] ^= out[i];
            out[i] ^= out[rev_i];
        }
    }

    int iterations = (int)log2(n);
    int M = 2;

    for (int i = 0; i < iterations; i++) {
        for (int j = 0; j < n; j += M) {
            int g = 0;
            for (int k = j; k < j + M/2; k++) {
                int U = out[k];
                int V = out[k + M/2] * omegas[g];
                out[k] = (U + V) % q;
                out[k + M/2] = (U - V + q) % q;
                g = (g + n / M) % n;
            }
        }
        M <<= 1;
    }

    return out;
}

int* cooley_tukey_intt(int* out, int q, int* omegas, int n, int invN) {
    out = cooley_tukey_ntt(out, q, omegas, n);

    // Scale output by 1/n
    for (int i = 0; i < n; i++) {
        out[i] = (out[i] * invN) % q;
    }

    return out;
}

void main() {
    int a[] = {1, 2, 3, 0, 0, 0, 0, 0};
    int mod = 17;
    int w[] = {1, 9, 13, 15, 16, 8, 4, 2};
    int n = 8;

    int* A = cooley_tukey_ntt(a, mod, w, n);
    printf("=====\n");
    for (int i = 0; i < n; i++) 
        printf("%d, ", A[i]);
    printf("\n");


    int invW[] = {1, 2, 4, 8, 16, 15, 13, 9};
    int invN = 15;

    int* B = cooley_tukey_intt(A, mod, invW, n, invN);

    for (int i = 0; i < n; i++) 
        printf("%d, ", B[i]);
    printf("\n");
}

File: SW/FNTT/fntt.py
import math

def reverse_bits(number, bit_length):
    # Reverses the bits of `number` up to `bit_length`.
    reversed = 0
    for i in range(0, bit_length):
        if (number >> i) & 1: 
            reversed |= 1 << (bit_length - 1 - i)
    return reversed

def cooley_tukey_ntt(out, q, omegas, n):
    # Reverses the bits of `number` up to `bit_length`.
    for i in range(n):
        rev_i = reverse_bits(i, n.bit_length() - 1)
        if rev_i > i:
            out[i] ^= out[rev_i]
            out[rev_i] ^= out[i]
            out[i] ^= out[rev_i]
    iterations = int(math.log2(n))

    M = 2
    for _ in range(iterations):
        for i in range(0, n, M):
            g = 0
            for j in range(0, M >> 1):
                k = i + j + (M >> 1)
                U = out[i + j]
                V = out[k] * omegas[g]
                out[i + j] = (U + V) % q
                out[k] = (U - V) % q
                g = g + n // M
        M <<= 1

    return out


def cooley_tukey_intt(out, q, omegas, n, invN):
    out = cooley_tukey_ntt(out, q, omegas, n)
    # Scale output by 1/n
    for i in range(n):
        out[i] = (out[i] * invN) % q

    return out

def demo():
    a = [0,1,2,3,4,5,6,7]
    mod = 29
    w = [1, 16, 24, 7, 25, 23, 20, 1]
    A = cooley_tukey_ntt(a,mod,w, 8)
    print("==NTT Result==")
    print(A)

    invW = [1, 2, 4, 8, 16, 15, 13, 9]
    invN = 15
    B = cooley_tukey_intt(A, mod, invW, 8, invN)

    print("==INTT Result==")
    print(B)

def main():
    demo()

if __name__ == '__main__':
    main()